pkg/controller/atlasprivateendpoint/atlasprivateendpoint_controller.go

```diff
@@ -0,0 +1,343 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/pointer"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/translation/privateendpoint"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/translation/project"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api"
+	akov2 "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/atlas"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/customresource"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/statushandler"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/workflow"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/indexer"
+)
+
+// AtlasPrivateEndpointReconciler reconciles a AtlasPrivateEndpoint object
+type AtlasPrivateEndpointReconciler struct {
+	Scheme                 *runtime.Scheme
+	Client                 client.Client
+	EventRecorder          record.EventRecorder
+	AtlasProvider          atlas.Provider
+	GlobalPredicates       []predicate.Predicate
+	Log                    *zap.SugaredLogger
+	privateEndpointService privateendpoint.PrivateEndpointService
+
+	ObjectDeletionProtection bool
+	independentSyncPeriod    time.Duration
+}
+
+// +kubebuilder:rbac:groups=atlas.mongodb.com,resources=atlasprivateendpoints,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=atlas.mongodb.com,resources=atlasprivateendpoints/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=atlas.mongodb.com,namespace=default,resources=atlasprivateendpoints,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=atlas.mongodb.com,namespace=default,resources=atlasprivateendpoints/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;patch
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch
+
+func (r *AtlasPrivateEndpointReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.Log.Infow("-> Starting AtlasPrivateEndpoint reconciliation")
+
+	akoPrivateEndpoint := akov2.AtlasPrivateEndpoint{}
+	result := customresource.PrepareResource(ctx, r.Client, req, &akoPrivateEndpoint, r.Log)
+	if !result.IsOk() {
+		return result.ReconcileResult(), errors.New(result.GetMessage())
+	}
+
+	return r.ensureCustomResource(ctx, &akoPrivateEndpoint)
+}
+
+func (r *AtlasPrivateEndpointReconciler) ensureCustomResource(ctx context.Context, akoPrivateEndpoint *akov2.AtlasPrivateEndpoint) (ctrl.Result, error) {
```

Comment:
I like this ensure separation, this regularizes the preamble in the following steps:
- **r.Reconcile** just unwraps the request into a full blown resource object.
- **r.ensure...** takes care of skipping, condition preparation, validation, checks for support and solves any parent reference. I wonder if we can make it generic.
- **r.handle...** evaluates the state machine.